File: src/control/ControlPictures.java
package control;

import java.awt.BorderLayout;
import java.awt.Component;
import java.awt.Dimension;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.io.File;

import javax.swing.BorderFactory;
import javax.swing.JButton;
import javax.swing.JComboBox;
import javax.swing.JPanel;
import javax.swing.JScrollPane;

import display.DisplayPictures;
import display.Scale;
import main.Settings;

/**
 * a {@code Control} for the Layer and Image Utility
 * @author McAJBen <McAJBen@gmail.com>
 * @since 2.0
 */
public class ControlPictures extends Control {

	private static final long serialVersionUID = -1679600820663944136L;
	
	/**
	 * the scroll menu of images inside the folder
	 */
	private PicturePanel pp;
	
	/**
	 * the folder which images are selected from
	 */
	private final File folder;
	
	/**
	 * the display to post images to
	 */
	private final DisplayPictures display;
	
	/**
	 * creates an instance of the {@code ControlPaint} class
	 * @param folder the folder that images are loaded from
	 * @param display the display to post images to
	 * @param allowList if more than 1 image should be allowed<br>
	 * - true will be for Layer Utility<br>
	 * - false will be for Image Utility
	 */
	public ControlPictures(File folder, DisplayPictures display, boolean allowList) {
		this.folder = folder;
		this.display = display;
		
		JPanel northPanel = getNorthPanel();
		
		JComboBox<Scale> scaleComboBox = new JComboBox<>(Scale.values());
		scaleComboBox.setBackground(Settings.CONTROL_BACKGROUND);
		scaleComboBox.setMaximumSize(new Dimension(100, 5000));
		scaleComboBox.addActionListener(new ActionListener() {
			public void actionPerformed(ActionEvent e) {
				display.setScaleMode(scaleComboBox.getSelectedItem());
			}
		});
		northPanel.add(scaleComboBox);
		
		JButton flipButton = Settings.createButton(Settings.ICON_FLIP);
		flipButton.setBackground(Settings.CONTROL_BACKGROUND);
		flipButton.addActionListener(new ActionListener() {
			public void actionPerformed(ActionEvent arg0) {
				display.flip();
			}
		});
		northPanel.add(flipButton);
		
		pp = new PicturePanel() {
			
			private static final long serialVersionUID = 2972394170217781329L;
			
			@Override
			protected void select(String name) {
				if (!allowList) {
					display.removeAllImages();
					for (Component c: getComponents()) {
						c.setBackground(Settings.DISABLE_COLOR);
					}
				}
				display.addImage(name);
			}

			@Override
			protected void deselect(String name) {
				display.removeImage(name);
			}
		};
		
		add(northPanel, BorderLayout.NORTH);
		
		JScrollPane jsp = new JScrollPane(pp);
		jsp.setBackground(Settings.CONTROL_BACKGROUND);
		jsp.setBorder(BorderFactory.createEmptyBorder());
		add(jsp, BorderLayout.CENTER);
		
		load();
		
		setVisible(true);
	}
	
	@Override
	public void setMainControl(boolean b) {
		if (b) {
			pp.rememberThumbnails(folder);
		}
		else {
			pp.forgetThumbnails();
		}
	}
	
	@Override
	protected void load() {
		if (folder.exists()) {
			pp.clearButtons();
			
			PPButtonCreator ppbc = new PPButtonCreator(pp, folder);
			ppbc.run();
			
			repaint();
			revalidate();
			display.removeAllImages();
		}
	}
}

File: src/main/kotlin/control/ControlPictures.kt
package control

import display.DisplayPictures
import display.Scale
import kotlinx.coroutines.async
import kotlinx.coroutines.runBlocking
import main.Mode
import util.*
import java.awt.BorderLayout
import java.awt.Dimension
import java.io.File
import javax.swing.BorderFactory
import javax.swing.JComboBox
import javax.swing.JScrollPane

/**
 * a `Control` for the Layer and Image Utility
 * @param display the display to post images to
 * @param mode which mode this panel is running
 * @author McAJBen@gmail.com
 */
class ControlPictures(
	private val display: DisplayPictures, mode: Mode
): Control() {

	companion object {
		private const val serialVersionUID = -1679600820663944136L
	}

	/**
	 * the folder that images are loaded from
	 */
	private val folder = Settings.getFolder(mode)

	/**
	 * data folder that contains cached thumbnails
	 */
	private val thumbnailFolder = Settings.getDataFolder(mode)

	/**
	 * if more than 1 image should be allowed
	 */
	private val allowList = mode == Mode.LAYER

	/**
	 * the scroll menu of images inside the folder
	 */
	private val picturePanel = object: PicturePanel(mode) {
		override fun onChange(button: PictureButton, isEnabled: Boolean) {
			if (isEnabled) {
				if (!allowList) {
					display.removeAllImages()
					buttons.forEach { it.setEnabled(false) }
				}
				display.addImage(button)
			} else {
				display.removeImage(button)
			}
		}
	}

	init {
		val scaleComboBox: JComboBox<Scale> = JComboBox<Scale>(Scale.values()).apply {
			background = Colors.CONTROL_BACKGROUND
			maximumSize = Dimension(
				100,
				5000
			)
			selectedItem = Scale.UP_SCALE
			addActionListener { display.setScaleMode(selectedItem as Scale) }
		}

		val flipButton = createButton(Resources.ICON_FLIP).apply {
			background = Colors.CONTROL_BACKGROUND
			addActionListener { display.flip() }
		}

		add(
			northPanel.apply {
				add(scaleComboBox)
				add(flipButton)
			},
			BorderLayout.NORTH
		)
		add(
			JScrollPane(picturePanel).apply {
				background = Colors.CONTROL_BACKGROUND
				border = BorderFactory.createEmptyBorder()
			},
			BorderLayout.CENTER
		)
		load()
		picturePanel.unloadButtons()
		isVisible = true
	}

	/**
	 * turns a thumbnail file into its normal file equal
	 * @return a file in the normal folders
	 */
	private fun File.thumbToFile(): File {
		return File(
			folder,
			name
		)
	}

	override fun setMainControl(b: Boolean) {
		if (b) {
			picturePanel.loadButtons()
		} else {
			picturePanel.unloadButtons()
		}
	}

	override fun load() {
		if (folder.exists()) {

			// remove any thumbnails for files that don't exist
			thumbnailFolder.listFiles()!!.filterNot {
				it.thumbToFile().exists()
			}.forEach {
				it.delete()
			}

			picturePanel.unloadButtons()

			runBlocking {
				folder.listFilesInOrder().filter {
					it.hasImageExtension()
				}.map {
					async { picturePanel.addPicture(it) }
				}.forEach {
					it.await()
				}
			}

			repaint()
			revalidate()
			display.removeAllImages()
			picturePanel.loadButtons()
		}
	}

	override fun onClosing() {}
}